File: repo_training_data/fix_html_entities.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// HTML entity replacement function
void decode_html_entities(const char* input, char* output) {
    const char* p = input;
    char* q = output;
    
    while (*p) {
        if (strncmp(p, "&amp;", 5) == 0) {
            *q++ = '&';
            p += 5;
        } else if (strncmp(p, "&lt;", 4) == 0) {
            *q++ = '<';
            p += 4;
        } else if (strncmp(p, "&gt;", 4) == 0) {
            *q++ = '>';
            p += 4;
        } else if (strncmp(p, "&quot;", 6) == 0) {
            *q++ = '"';
            p += 6;
        } else if (strncmp(p, "&#39;", 5) == 0) {
            *q++ = '\'';
            p += 5;
        } else {
            *q++ = *p++;
        }
    }
    *q = '\0';
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <file>\n", argv[0]);
        return 1;
    }
    
    const char* filename = argv[1];
    
    // Read entire file
    FILE* f = fopen(filename, "r");
    if (!f) {
        perror("fopen");
        return 1;
    }
    
    fseek(f, 0, SEEK_END);
    long size = ftell(f);
    fseek(f, 0, SEEK_SET);
    
    char* content = malloc(size + 1);
    if (!content) {
        fclose(f);
        fprintf(stderr, "malloc failed\n");
        return 1;
    }
    
    fread(content, 1, size, f);
    content[size] = '\0';
    fclose(f);
    
    // Decode entities
    char* decoded = malloc(size + 1);
    if (!decoded) {
        free(content);
        fprintf(stderr, "malloc failed\n");
        return 1;
    }
    
    decode_html_entities(content, decoded);
    
    // Write back
    f = fopen(filename, "w");
    if (!f) {
        free(content);
        free(decoded);
        perror("fopen");
        return 1;
    }
    
    fwrite(decoded, 1, strlen(decoded), f);
    fclose(f);
    
    free(content);
    free(decoded);
    
    printf("Fixed HTML entities in %s\n", filename);
    return 0;
}

File: archive/legacy_code/tools_old/fix_html_entities.py
#!/usr/bin/env python3
"""
HTML Entity Fixer Tool

This tool fixes HTML entity encoding issues that sometimes occur when
creating files through the AI assistant interface.

Common entities fixed:
- &amp; -> &
- &lt; -> <
- &gt; -> >
- &quot; -> "
- &#39; -> '

Usage:
    python3 tools/fix_html_entities.py <file>
    python3 tools/fix_html_entities.py src/ai/cllm_threads.c
"""

import sys
import os

def decode_html_entities(content):
    """Decode HTML entities in content."""
    replacements = {
        '&amp;': '&',
        '&lt;': '<',
        '&gt;': '>',
        '&quot;': '"',
        '&#39;': "'",
    }
    
    for entity, char in replacements.items():
        content = content.replace(entity, char)
    
    return content

def fix_file(filename):
    """Fix HTML entities in a file."""
    if not os.path.exists(filename):
        print(f"Error: File '{filename}' not found")
        return 1
    
    # Read file
    try:
        with open(filename, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception as e:
        print(f"Error reading file: {e}")
        return 1
    
    # Check if file has HTML entities
    has_entities = any(entity in content for entity in ['&amp;', '&lt;', '&gt;', '&quot;', '&#39;'])
    
    if not has_entities:
        print(f"No HTML entities found in {filename}")
        return 0
    
    # Decode entities
    decoded = decode_html_entities(content)
    
    # Write back
    try:
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(decoded)
        print(f"✓ Fixed HTML entities in {filename}")
        return 0
    except Exception as e:
        print(f"Error writing file: {e}")
        return 1

def main():
    if len(sys.argv) != 2:
        print("Usage: python3 tools/fix_html_entities.py <file>")
        print("\nExample:")
        print("  python3 tools/fix_html_entities.py src/ai/cllm_threads.c")
        return 1
    
    filename = sys.argv[1]
    return fix_file(filename)

if __name__ == '__main__':
    sys.exit(main())